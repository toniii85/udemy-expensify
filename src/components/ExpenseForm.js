import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";

const now = moment();

class ExpenseForm extends React.Component {
  constructor(props){
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense ? (props.expense.amount / 100) : "",
      createdAt: props.expense ? moment(props.expense.createdAt): moment(),
      calendarFocused: false,
      error: "",
    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      //Set error Please provide description and amount
      this.setState(() => ({ error: "Por favor proporciona una descripcion y un importe" }));
    } else {
      //Clear error
      this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note,
      });
    }
  }
  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Descripción"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text"
            placeholder="Importe"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Añade una nota para el gasto (opcional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          ></textarea>
          <button>Añadir Gasto</button>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;