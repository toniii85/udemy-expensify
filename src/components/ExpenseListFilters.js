import React from "react";
import { connect } from "react-redux";
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from "../actions/filters";
import { DateRangePicker } from "react-dates";

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null,
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  }
  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  }
  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };
  onSortChange = (e) => {
    if (e.target.value === "amount") {
      this.props.sortByAmount();
    } else {
      this.props.sortByDate();
    }
  }
  render() {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input
              className="text-input"
              placeholder="BÃºsqueda de gastos"
              type="text"
              value={this.props.filters.text}
              onChange={this.onTextChange} />
          </div>
          <div className="input-group__item">
            <select
              className="select"
              value={this.props.filters.sortBy}
              onChange={this.onSortChange}>
              <option value="date" >Fecha</option>
              <option value="amount" >Importe</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker
              startDate={this.props.filters.startDate}
              startDatePlaceholderText={"Fecha inicio"}
              endDate={this.props.filters.endDate}
              endDatePlaceholderText={"Fecha fin"}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.calendarFocused}
              showClearDates={true}
              numberOfMonths={1}
              onFocusChange={this.onFocusChange}
              isOutsideRange={() => false}
            />
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  filters: state.filters,
});

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate)),
});
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);