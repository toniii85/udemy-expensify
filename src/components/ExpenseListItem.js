import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";
import numeral from "numeral";

numeral.register("locale", "es", {
  delimiters: {
      thousands: " ",
      decimal: ","
  },
  abbreviations: {
      thousand: "k",
      million: "m",
      billion: "b",
      trillion: "t"
  },
  ordinal : function (number) {
      return number === 1 ? "er" : "o";
  },
  currency: {
      symbol: "â‚¬"
  }
});
numeral.locale("es");

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
  <div>
    <Link to={`/edit/${id}`}>
      <h3>{description}</h3>
    </Link>
    <p>
      {numeral(amount / 100).format("0,0[.]00 $")}
      -
      {moment(createdAt).format("DD [de] MMMM YYYY")}
    </p>
  </div>
);

export default ExpenseListItem;