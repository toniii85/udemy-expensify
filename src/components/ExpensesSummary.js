import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import numeral from "numeral";
import selectExpenses from "../selectors/expenses";
import selectExpensesTotal from "../selectors/expenses-total";


export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
  const expenseWord = expenseCount === 1 ? "gasto" : "gastos";
  const formattedExpensesTotal = numeral(expensesTotal / 100).format("0,0[.]00 $");
  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">Viendo <span>{expenseCount}</span> {expenseWord}, el total es:  <span>{formattedExpensesTotal}</span></h1>
        <div className="page-header__actions">
          <Link className="button" to="/create">AÃ±adir gasto</Link>
        </div>
      </div>
    </div>
  );
}

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: selectExpensesTotal(visibleExpenses),
  };
};
export default connect(mapStateToProps)(ExpensesSummary);