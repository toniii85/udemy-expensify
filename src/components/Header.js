import React from "react";
import { NavLink } from "react-router-dom";

const Header = () => (
  <header>
    <h1>Expensify</h1>
    <NavLink to="/" activeClassName="is-active" exact>Panel</NavLink>
    <NavLink to="/create" activeClassName="is-active" >Crear gasto</NavLink>
    <NavLink to="/help" activeClassName="is-active" >Ayuda</NavLink>
  </header>
);

export default Header;