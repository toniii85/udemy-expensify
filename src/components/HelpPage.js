import React from "react";


const HelpPage = () => (
  <div>
    This is from help expense component
  </div>
);


export default HelpPage;