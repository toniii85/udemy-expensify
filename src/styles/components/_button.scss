.button {
  background: $blue;
  color: white;
  border: none;
  padding: $s-size;
  font-weight: 300;
  font-size: $font-size-large;
  display: inline-block;
  text-decoration: none;
  line-height: 1;

  // height: $l-size;
  // width: 15rem;
  // border-radius: 2px;
  // font-weight: 200;
  // font-size: $m-size;
}

.button--link {
  background: none;
}

.button--secondary {
  background: $soft-grey;
}