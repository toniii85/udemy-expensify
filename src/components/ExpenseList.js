import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";

export const ExpenseList = (props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Gastos</div>
      <div className="show-for-desktop">Gasto</div>
      <div className="show-for-desktop">Importe</div>
    </div>
    <div className="list-body">
      {
        props.expenses.length === 0 ? (
          <div className="list-item list-item--message">
            <p>No hay gastos</p>
          </div>
        ) : (
            props.expenses.map((expense) => (
              <ExpenseListItem key={expense.id} {...expense} />
            ))
          )
      }
    </div>
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters),
  };
};

export default connect(mapStateToProps)(ExpenseList);