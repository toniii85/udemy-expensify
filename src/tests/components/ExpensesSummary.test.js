import React from "react";
import { shallow } from "enzyme";
import { ExpensesSummary } from "../../components/ExpensesSummary";
import numeralLocale from "../../config/numeral-locale";

numeralLocale();

test("should correctly render expeneses summary with 1 expense", () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={235} />);

  expect(wrapper).toMatchSnapshot();
});

test("should correctly render expeneses summary with multiple expense", () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={23} expensesTotal={3243222} />);
  expect(wrapper).toMatchSnapshot();
});