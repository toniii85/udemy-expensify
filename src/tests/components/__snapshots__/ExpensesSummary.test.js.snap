// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render expeneses summary with 1 expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viendo 
      <span>
        1
      </span>
       
      gasto
      , el total es:  
      <span>
        2,35 €
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Añadir gasto
      </Link>
    </div>
  </div>
</div>
`;

exports[`should correctly render expeneses summary with multiple expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viendo 
      <span>
        23
      </span>
       
      gastos
      , el total es:  
      <span>
        32 432,22 €
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Añadir gasto
      </Link>
    </div>
  </div>
</div>
`;
